Add Android Studio support

m4gr3d opened the pull request:
This provides the necessary metadata (e.g: build.gradle, CMakeList, etc...) to use Android Studio as IDE to work on the Godot engine.

This is especially relevant when working on Godot Android related code since Android Studio is one of the few IDE that provides editor support for java and c++ code.

Calinou commented:
Including a `CMakeLists.txt` file might induce people in error, thinking they can build the engine using CMake (which isn't the case).

m4gr3d commented:
> Including a `CMakeLists.txt` file might induce people in error, thinking they can build the engine using CMake (which isn't the case).

The `CMakeLists.txt` is only intended to be used by Android Studio to index the project c++ files. 
Would adding a comment to the file, to specify it's not used for building the engine, help?

reduz commented:
This is pretty good, but I need to do a few changes to Android build system the coming days to make exporting easier, so will put this one on hold for a bit until I'm done.

akien-mga commented:
I'm against adding so many files to the root directory, we've already refused similar IDE-specific files in the past. If we accepted them all we'd have 100 files in the root directory and no visibility on what files are actually important.

Can this be in a subfolder, e.g. `misc/ide/android_studio`?

m4gr3d commented:
> I'm against adding so many files to the root directory, we've already refused similar IDE-specific files in the past. If we accepted them all we'd have 100 files in the root directory and no visibility on what files are actually important.
> 
> Can this be in a subfolder, e.g. `misc/ide/android_studio`?

Sounds good. I'll update the PR to move these IDE-specific files in a subfolder as you proposed.

m4gr3d commented:
@akien-mga I've updated the PR and moved the gradle build files into `misc/ide/android_studio` as suggested. The PR is ready for review.

shartte commented:
This would make it possible to use Clion for editing too, right?
If one is not interested in the Android part of it.

m4gr3d commented:
> This would make it possible to use Clion for editing too, right?
> If one is not interested in the Android part of it.

@shartte Yes! I tested it and by pointing to the `CMakeLists.txt` file, you can get CLion to support editing of the Godot project. 
Once the project is loaded, you can even [update the project root](https://www.jetbrains.com/help/clion/change-project-root-directory.html) within CLion to have a better view of the project directory structure.

I'll rename `misc/ide/android_studio` to `misc/ide/jetbrains`  to remove the implication this is only usable by Android Studio.

reduz commented:
Please wait for #27781 to be merged.

Calinou commented:
@reduz #27781 has been merged, could this one be merged as well (after conflicts are resolved)?

Deliner commented:
@reduz is it time to merge?

vnen commented:
With this it's possible to run a Godot project in the phone from Android Studio? If so, it would help a lot in debugging.

m4gr3d commented:
> With this it's possible to run a Godot project in the phone from Android Studio? If so, it would help a lot in debugging.

@vnen In short, yes!

The added gradle build file allows to generate and export a Godot android library artifact (aar file) which can be linked to a regular android project to run a Godot project from Android Studio.

I've been using it for my own development and testing; there's still some manual steps involved (e.g: exporting the Godot project in a zip file, and unpacking it in the Android project assets directory), but I'm hoping to clean and automate the process in the near future.

Regarding debugging, while I'm able to debug Godot Android java classes, I'm still running into issues debugging the C++ classes, but I'm hoping that to just be a debugging symbol configuration issue.

akien-mga commented on misc/ide/jetbrains/settings.gradle:
```diff
@@ -0,0 +1,2 @@
+// Configure the root project.
+rootProject.name = "Godot"
```
Should have a newline character at end of file.

akien-mga commented on misc/ide/jetbrains/build.gradle:
```diff
@@ -0,0 +1,108 @@
+
+def pathToRootDir = "../../../"
+// Note: Only keep the abis you support to speed up the gradle 'assemble' task.
+def supportedAbis = ["armv7", "arm64v8", "x86"]
```
Why not `x86_64`?

m4gr3d replied:
I originally went by the documentation [here](https://docs.godotengine.org/en/latest/development/compiling/compiling_for_android.html#adding-support-for-x86-devices), hence only these three since those are the ones mentioned there.

The `godot/platform/android/SCub` build file shows support for `armv6` and `x86_64` as well, so I'll add those as well.

m4gr3d replied:
Btw, the `armv6` build support seems to be broken. I get the following error messages when attempting to build it:
```
/var/folders/hf/nggb0mq11hv4t5s0vzfvs_8m00bx2l/T/thread_jandroid-2bb15c.s: Assembler messages:
/var/folders/hf/nggb0mq11hv4t5s0vzfvs_8m00bx2l/T/thread_jandroid-2bb15c.s:119: Error: selected processor does not support `ldrexd r2,r3,[r0]' in ARM mode
/var/folders/hf/nggb0mq11hv4t5s0vzfvs_8m00bx2l/T/thread_jandroid-2bb15c.s:122: Error: selected processor does not support `strexd r1,r6,r7,[r0]' in ARM mode
/var/folders/hf/nggb0mq11hv4t5s0vzfvs_8m00bx2l/T/thread_jandroid-2bb15c.s:322: Error: selected processor does not support `ldrexd r2,r3,[r0]' in ARM mode
/var/folders/hf/nggb0mq11hv4t5s0vzfvs_8m00bx2l/T/thread_jandroid-2bb15c.s:325: Error: selected processor does not support `strexd r1,r4,r5,[r0]' in ARM mode
```

Seems as far as [Aug 2017](https://github.com/godotengine/godot/issues/10711) it was recommended to use `armv7` instead. Should we consider `armv6` to be deprecated?

akien-mga commented:
Seems good to me overall, but the commits will need to be squashed together. Added a couple comments too.

m4gr3d committed:
Add Android Studio support.

m4gr3d commented:
> Seems good to me overall, but the commits will need to be squashed together. Added a couple comments too.

@akien-mga I've addressed your comments. The PR is ready for review.

akien-mga commented:
Thanks!

akien-mga commented:
It would be great to add documentation on how to use these files in https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide.html

m4gr3d commented:
> It would be great to add documentation on how to use these files in https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide.html

@akien-mga Sounds good. I'll prep a PR to update the documentation as well.

grusad commented:
Having trouble compiling using Clion on the current master branch. Error: platform_config.h: No such file or directory #include "platform_config.h".

DDL-Blue commented:
Hi,
why can't I see the misc/ide folder in master branch, even though this PR was merged into master? Is it just my fault, or did something go wrong with this PR?

m4gr3d commented:
@DDL-Blue The directory structure has been updated and the `misc/ide` folder is no longer needed.
You can find the updated setup instructions [here](https://docs.godotengine.org/en/latest/development/cpp/configuring_an_ide.html#doc-configuring-an-ide-android-studio).

DDL-Blue commented:
@m4gr3d Eh, ok. The thing is that it also contained  misc/ide/jetbrains/CMakeLists.txt file that is needed when editing the project using CLion. This is why I was looking for it in a first place. And also why it was linked in [this issue](https://github.com/godotengine/godot/issues/29596) as a solution (which it is no more). The misc/ide folder did seem like a great solution for this. The file is not in a root folder, so it prevents the confusion Calinou mentioned in his first comment and CLion is a quite popular IDE.

DDL-Blue commented:
@m4gr3d Oh, I've found the file in the platform/android/java/lib/CMakeLists.txt folder. It should be noted in the documentation as well.

m4gr3d commented:
Good point, will do.

m4gr3d commented:
@DDL-Blue just saw your updated comment. Let me know if you need assistance to update the docs.

DDL-Blue commented:
@m4gr3d It is [PR 3021](https://github.com/godotengine/godot-docs/pull/3021) in godot-docs repo.

LeeTeng2001 commented:
Hi, I would love to use cmakelist for my jetbrain ide intellisense but I couldn't seem to find cmakelist.txt in the project, where did it go ? thank you

LeeTeng2001 commented:
My bad! I found it
For people that run into the same problem with me and found this thread, all the information is located inside https://docs.godotengine.org/en/stable/development/cpp/configuring_an_ide/clion.html
